src/GraphQL/Validation/Rules.Custom/InputFieldsAndArgumentsOfCorrectLength.cs

```diff
@@ -0,0 +1,81 @@
+using System.Threading.Tasks;
+using GraphQL.Execution;
+using GraphQL.Language.AST;
+using GraphQL.Types;
+using GraphQL.Validation.Errors;
+
+namespace GraphQL.Validation.Rules
+{
+    /// <summary>
+    /// Validation rule that checks minimum and maximum length of provided values for input fields and
+    /// arguments that marked with <see cref="LengthDirective"/> directive. Doesn't check default values.
+    ///
+    /// <br/><br/>
+    /// This is not a standard validation rule that is not in the specification.
+    /// </summary>
+    public class InputFieldsAndArgumentsOfCorrectLength : IValidationRule
+    {
+        /// <summary>
+        /// Returns a static instance of this validation rule.
+        /// </summary>
+        public static readonly InputFieldsAndArgumentsOfCorrectLength Instance = new InputFieldsAndArgumentsOfCorrectLength();
+
+        /// <inheritdoc/>
+        /// <exception cref="InputFieldsAndArgumentsOfCorrectLengthError"/>
+        public Task<INodeVisitor> ValidateAsync(ValidationContext context) => _nodeVisitor;
+
+        private static readonly Task<INodeVisitor> _nodeVisitor = new NodeVisitors(
+            new MatchingNodeVisitor<Argument>((arg, context) => CheckLength(arg, context.TypeInfo.GetArgument(), context)),
+            new MatchingNodeVisitor<ObjectField>((field, context) =>
+            {
+                if (context.TypeInfo.GetInputType(1) is InputObjectGraphType input)
+                    CheckLength(field, input.GetField(field.Name), context);
+            })
+        ).ToTask();
+
+        private static void CheckLength(IHaveValue node, IProvideMetadata provider, ValidationContext context)
+        {
+            if (provider == null || !provider.HasAppliedDirectives())
+                return;
+
+            var lengthDirective = provider.GetAppliedDirectives().Find("length");
+            if (lengthDirective == null)
+                return;
+
+            var min = lengthDirective.FindArgument("min")?.Value;
+            var max = lengthDirective.FindArgument("max")?.Value;
+
+            if (node.Value is NullValue)
+            {
+                if (min != null)
+                    context.ReportError(new InputFieldsAndArgumentsOfCorrectLengthError(context, node, null, (int?)min, (int?)max));
+            }
+            else if (node.Value is StringValue strLiteral)
+            {
+                CheckStringLength(strLiteral.Value);
+            }
+            else if (node.Value is VariableReference vRef && context.Inputs != null)
+            {
+                var operation = string.IsNullOrWhiteSpace(context.OperationName)
+                    ? context.Document.Operations.FirstOrDefault()
+                    : context.Document.Operations.WithName(context.OperationName);
+
+                //TODO: change to get single variable
+                var values = ExecutionHelper.GetVariableValues(context.Document, context.Schema, operation?.Variables, context.Inputs);
```

Comment:
This looks like bad practice.  We should never call `GetVariableValues` within a validation rule.  It would have to parse all variables a second time.  Maybe multiple times depending on how many times this ran.  Besides, does this even work in practice?  Assuming there is a length directive on a string as part of an input object, doesn't this get skipped because there is no directive applied to the input argument?  At first glance it seems like it would only work if an argument type was a `StringGraphType`.

It would seem that we need to design a validation framework for variables, which for speed purposes should be processed at the same time that the variables are being parsed.

Reply:
>  It would have to parse all variables a second time. 

I know.

> Besides, does this even work in practice? Assuming there is a length directive on a string as part of an input object, doesn't this get skipped because there is no directive applied to the input argument? 

It works. See `HasAppliedDirectives` / `GetAppliedDirectives().Find` above.

> it's possible that we would need 3 separate validation passes

Something like this. Not a small job. Not my goal now.

Reply:
> design a validation framework for variables

See my first steps with `IVariableVisitor`.

Reply:
Now that we have a `IVariableVisitor` framework, it is at least possible to provide a fully working sample.  I would like to review further.

Reply:
Back to this particular line of code ... this would only run if there was a length directive applied to a string type that is a variable reference, correct?  So as long as we can make it only run `GetVariableValue` to retrieve only the variable desired, it should not be unreasonably slower even for the most complex of queries/mutations.  This should be implemented before this PR is merged.  (It is of course possible to manually pull the string value from the inputs, but there is no good reason to do so.)

Reply:
Once everything is done we can write some complex mutations and see the performance impact, as well as verifying that there is little/no performance impact for code without directives.

Reply:
> this would only run if there was a length directive applied to a string type that is a variable reference, correct?

Yes.

Reply:
> it's possible that we would need 3 separate validation passes

Done. As a result, I came to the conclusion that it would be logical to build variables at the validation stage. Before my changes, they were created when building ExecutionContext right afted validation stage.  A side effect is now more validation rule triggers due to missing variables or setting them incorrectly. I had to edit a lot of tests. Variable errors were previously thrown as exceptions after direct validation. I just changed `throw` to `context.ReportError` inside `GetVariableValues`. I believe that overall the code has become better and more logical, despite the fact that now, at the stage of validation, new returned errors seem to be similar to the existing ones. Look at the tests, it will be clear there.